pkg/proxy/ipvs/meta-proxier.go

```diff
@@ -0,0 +1,193 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipvs
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/proxy"
+	utilnet "k8s.io/utils/net"
+)
+
+type MetaProxier struct {
+	ipv4Proxier proxy.ProxyProvider
+	ipv6Proxier proxy.ProxyProvider
+}
+
+type ipFamily int
+
+const (
+	familyUnknown = iota + 1
+	familyIpv4
+	familyIpv6
+)
+
+// NewMetaProxier returns a dual-stack "meta-proxier". Proxier API
+// calls will be dispatched to the ProxyProvider instances depending
+// on address family.
+func NewMetaProxier(ipv4Proxier, ipv6Proxier proxy.ProxyProvider) *MetaProxier {
+	return &MetaProxier{
+		ipv4Proxier: ipv4Proxier,
+		ipv6Proxier: ipv6Proxier,
+	}
+}
+
+// Sync immediately synchronizes the ProxyProvider's current state to
+// proxy rules.
+func (proxier *MetaProxier) Sync() {
+	proxier.ipv4Proxier.Sync()
+	proxier.ipv6Proxier.Sync()
+}
+
+// SyncLoop runs periodic work.  This is expected to run as a
+// goroutine or as the main loop of the app.  It does not return.
+func (proxier *MetaProxier) SyncLoop() {
+	go proxier.ipv6Proxier.SyncLoop() // Use go-routine here!
+	proxier.ipv4Proxier.SyncLoop()    // never returns
+}
+
+// OnServiceAdd is called whenever creation of new service object is observed.
+func (proxier *MetaProxier) OnServiceAdd(service *v1.Service) {
+	family := proxier.serviceFamily(service)
+	switch family {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnServiceAdd(service)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnServiceAdd(service)
+	}
+}
+
+// OnServiceUpdate is called whenever modification of an existing
+// service object is observed.
+func (proxier *MetaProxier) OnServiceUpdate(oldService, service *v1.Service) {
+	family := proxier.serviceFamily(service)
+	if family == familyUnknown {
+		family = proxier.serviceFamily(oldService)
+	}
+	switch family {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnServiceUpdate(oldService, service)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnServiceUpdate(oldService, service)
+	}
+}
+
+// OnServiceDelete is called whenever deletion of an existing service
+// object is observed.
+func (proxier *MetaProxier) OnServiceDelete(service *v1.Service) {
+	switch proxier.serviceFamily(service) {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnServiceDelete(service)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnServiceDelete(service)
+	}
+}
+
+// OnServiceSynced is called once all the initial event handlers were
+// called and the state is fully propagated to local cache.
+func (proxier *MetaProxier) OnServiceSynced() {
+	proxier.ipv4Proxier.OnServiceSynced()
+	proxier.ipv6Proxier.OnServiceSynced()
+}
+
+func (proxier *MetaProxier) serviceFamily(service *v1.Service) ipFamily {
```

Comment:
does this get easier since we added an ipFamily field in #79386

Reply:
There is one `IPFamily` but the problem is that the normal value is "Default" and `kube-proxy` does not know the default family, please see [slack](https://kubernetes.slack.com/archives/CGF3A900N/p1565812069098000).